banyan/src/zstd_dag_cbor_seq.rs

```diff
@@ -248,42 +249,40 @@ impl ZstdDagCborSeq {
     }
 
     /// encrypt using the given key and nonce
-    pub fn encrypt(
-        &self,
-        nonce: &chacha20::XNonce,
-        key: &chacha20::Key,
-    ) -> anyhow::Result<Vec<u8>> {
-        self.clone().into_encrypted(nonce, key)
+    pub fn encrypt(&self, key: &chacha20::Key, mut offset: u64) -> anyhow::Result<Vec<u8>> {
+        self.clone().into_encrypted(key, &mut offset)
     }
 
     /// convert into an encrypted blob, using the given key and nonce
-    pub fn into_encrypted(
-        self,
-        nonce: &chacha20::XNonce,
-        key: &chacha20::Key,
-    ) -> anyhow::Result<Vec<u8>> {
+    pub fn into_encrypted(self, key: &chacha20::Key, offset: &mut u64) -> anyhow::Result<Vec<u8>> {
         let Self { mut data, links } = self;
         // encrypt in place with the key and nonce
-        let mut chacha20 = XChaCha20::new(key, nonce);
+        let mut chacha20 = XChaCha20::new(key, &NONCE.into());
+        let len = data.len();
+        chacha20.seek(*offset);
         chacha20.apply_keystream(&mut data);
         // add the nonce
-        data.extend(nonce.as_slice());
+        data.extend(&offset.to_be_bytes());
+        *offset += len as u64;
         // encode via IpldNode
         let result = DagCborCodec.encode(&IpldNode::new(links, data))?;
         Ok(result)
     }
 
     /// decrypt using the given key
-    pub fn decrypt(data: &[u8], key: &chacha20::Key) -> anyhow::Result<Self> {
+    pub fn decrypt(data: &[u8], key: &chacha20::Key) -> anyhow::Result<(Self, u64)> {
         let (links, mut encrypted) = DagCborCodec.decode::<IpldNode>(data)?.into_data()?;
         let len = encrypted.len();
         anyhow::ensure!(len >= EXTRA_LEN);
-        let (compressed, nonce) = encrypted.split_at_mut(len - EXTRA_LEN);
-        XChaCha20::new(key, (&*nonce).into()).apply_keystream(compressed);
+        let (compressed, offset) = encrypted.split_at_mut(len - EXTRA_LEN);
```

Comment:
Why is this necessary? The offset should be known implicitly due to the position in the tree, right?

Reply:
If you would never remove anything from the tree, you could infer the offset from the position in the tree. But we do want to have things like retain, which can forget parts of the tree. For that you need to explicitly store the offsets.

Now you could come up with a scheme where you only sometimes need the offset, but that would be rather complicated. Also, I prefer the blocks to be self-contained, so to look at a block you just need the secrets.